Persistent admin login

SupaHotBall opened the pull request:
## Summary

Persistent login system for the Admin section using JWT authentication. After successful login, the token is stored in `localStorage`, allowing the session to persist even after a page refresh.

## Changes Implemented

-  Fetched token from backend upon successful login
-  Stored JWT token in `localStorage`
-  Checked for existing token on Admin page load to auto-login
-  Displayed Admin Dashboard when token is valid
-  Used `localStorage.getItem("token")` to maintain session state

## Files Modified

- `LoginForm.jsx`:  
  - Changed to fetch from backend `/login`
  - Saved token and user role in `localStorage`

- `AdminPage.jsx`:  
  - Checked `localStorage` on load for existing token
  - Automatically authenticated if token exists

SupaHotBall committed:
Feat: Implement persistent Admin login using JWT and localStorage

squigggggle commented on Frontend/src/components/LoginForm.jsx:
```diff
@@ -6,15 +6,26 @@ const LoginForm = ({ onLoginSuccess }) => {
-      setError('Invalid credentials. Please try again.');
+
+    try {
+      const response = await fetch('http://127.0.0.1:5000/login', {
```
This will work when you are developing locally, but will not work when the app is deployed.
One option would be to have an `.env` variable that stores the url, or we need to separate the fetching logic into another component that has the url defined within that, like we have been doing in advanced app dev.

squigggggle commented on Frontend/src/components/LoginForm.jsx:
```diff
@@ -6,15 +6,26 @@ const LoginForm = ({ onLoginSuccess }) => {
+      const data = await response.json();
+
+      if (response.ok && data.token) {
+        localStorage.setItem('token', data.token); // Save the JWT
```
The token should start with `Bearer ` in front of the actual token.

squigggggle committed:
fix: update requirements with correct packages

squigggggle committed:
feat: .env used for fetch url

squigggggle commented on backend/requirements.txt:
I tried to run everything locally but there were missing packages, so I've updated them here

squigggggle commented on Frontend/src/components/LoginForm.jsx:
```diff
@@ -6,15 +6,29 @@ const LoginForm = ({ onLoginSuccess }) => {
-      setError('Invalid credentials. Please try again.');
+
+    try {
+      const response = await fetch(`${import.meta.env.VITE_API_URL}/login`, {
```
I've changed the hardcoded local hosted url to an `,env` instead, you can create an `,env` in the frontend and put the url after `VITE_API_URL`

squigggggle commented on Frontend/src/components/AdminPage.jsx:
```diff
@@ -1,18 +1,32 @@
+  const handleScrapeStart = async () => {
+    const token = localStorage.getItem('token');
+
+    const response = await fetch('http://127.0.0.1:5000/run-spiders', {
```
This will need changed to the `VITE_API_URL` since I have changed it in the login form

squigggggle commented on Frontend/src/components/AdminPage.jsx:
```diff
@@ -1,18 +1,32 @@
+
+    const response = await fetch('http://127.0.0.1:5000/run-spiders', {
+      headers: {
+        Authorization: `Bearer ${token}`,
```
This will not work as it is not being stored as `Bearer <token>` in the login form

squigggggle commented on Frontend/src/components/AdminPage.jsx:
The websocket for the scraper information is also missing, should be easy to add if you refer back to the code on main

squigggggle commented on Frontend/src/components/AdminPage.jsx:
```diff
@@ -1,18 +1,32 @@
+  // Check token on mount
+  useEffect(() => {
+    const token = localStorage.getItem('token');
+    if (token) setIsAuthenticated(true);
```
I don't think this checks whether the token is correct or not timed out, just that it exists. The user should be denied access to the admin page if the token hasn't been sent to the backend to be checked.

squigggggle reviewed (changes requested):
I have left some comments on parts that I think still need some work.
I think its good that the login page has been removed in place of just putting it on the admin page instead, one less page to worry about when its only going to be an admin user logging into the website anyway.

SupaHotBall committed:
Fix: Updated login and admin components to support JWT with Bearer token, environment variables for API URL, and token verification with backend

SupaHotBall commented:
### What’s Changed

- Refactored `LoginForm.jsx`:
  - Replaced hardcoded credentials with form inputs.
  - Fetched login token from backend via `.env` API URL.
  - Stored token in `localStorage` with `Bearer` prefix.

- Refactored `AdminPage.jsx`:
  - Verified token by calling `/admin` route on backend.
  - Used `.env` variable (`VITE_API_URL`) instead of hardcoded `localhost`.
  - Added proper Authorization header when calling protected route `/run-spiders`.

SupaHotBall committed:
Fix: verifying the token from the localstore

squigggggle reviewed (approved):
Yes!!!!!!!!!! Now the admin page login is persistent when you reload or navigate to another page !!!!!!!!!!!!!!!!!!!!!!!!!!!!! 👍👍👍👍👍👍👍👍👍👍👍👍👍

squigggggle committed:
fix: broken package-lock

squigggggle committed:
fix: generate new package-lock.json

squigggggle committed:
fix: package from staging

squigggggle committed:
Merge branch 'staging' into preserve-admin-login

squigggggle committed:
fix: reinstall recharts after resetting packages
